### PR TITLE
Add X-CLIP

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [X-CLIP](https://github.com/microsoft/VideoX/tree/master/X-CLIP), which is a minimal extension of CLIP for video-language pre-training.

To do:

- [x] upload all checkpoints to the hub, as part of the `microsoft` organization

### CODE REVIEW

**Comment by NielsRogge** [comment]
Many tests fail due to the following error:

> ModuleNotFoundError: No module named 'transformers.models.xclip'

This is probably because I first called the model folder "xclip", which is now called "x_clip". Still, wondering why it keeps looking for the module models.clip. If anyone has any pointers, that would be greatly appreciated.

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -367,6 +368,13 @@
         "WAVLM_PRETRAINED_CONFIG_ARCHIVE_MAP",
         "WavLMConfig",
     ],
+    "models.x_clip": [
+        "X_CLIP_PRETRAINED_CONFIG_ARCHIVE_MAP",
```
Comment (sgugger) [inline src/transformers/__init__.py:372]: For consistency of names, this should be `XCLIP_PRETRAINED_CONFIG_ARCHIVE_MAP` here.

File: src/transformers/models/auto/tokenization_auto.py [hunk 2]
```diff
@@ -253,6 +253,13 @@
             ("wav2vec2", ("Wav2Vec2CTCTokenizer", None)),
             ("wav2vec2-conformer", ("Wav2Vec2CTCTokenizer", None)),
             ("wav2vec2_phoneme", ("Wav2Vec2PhonemeCTCTokenizer", None)),
+            (
+                "x-clip",
```
Comment (sgugger) [inline src/transformers/models/auto/tokenization_auto.py:257]:
```suggestion
                "xclip",
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 3]
```diff
@@ -0,0 +1,1486 @@
...
+        causal_attention_mask: torch.Tensor,
+        output_attentions: Optional[bool] = False,
+    ) -> Tuple[torch.FloatTensor]:
+        """
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`): attention mask of size
+                `(batch, 1, tgt_len, src_len)` where padding elements are indicated by very large negative values.
+                `(config.encoder_attention_heads,)`.
+            output_attentions (`bool`, *optional*):
```
Comment (sgugger) [inline src/transformers/models/x_clip/modeling_x_clip.py:427]: causal_attention_mask is missing here.

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 4]
```diff
@@ -0,0 +1,1486 @@
...
+    def _shape(self, tensor: torch.Tensor, seq_len: int, bsz: int):
+        return tensor.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2).contiguous()
+
+    def forward(self, queries, keys, values):
+        """Input shape: Batch x Time x Channel"""
+        B, N, C = queries.shape
+        B, M, C = keys.shape
+        q = self.q_proj(queries).reshape(B, N, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        k = self.k_proj(keys).reshape(B, M, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        v = self.v_proj(values).reshape(B, M, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
```
Comment (sgugger) [inline src/transformers/models/x_clip/modeling_x_clip.py:1159]: Here please use better variable names for `B`, `M`, `N`, `C`, `q`, `k` and `v`.

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 5]
```diff
@@ -0,0 +1,1486 @@
...
+            pixel_values=pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        pooled_output = vision_outputs[1]  # pooled_output
+        image_features = self.visual_projection(pooled_output)
+
+        # TODO add the following:
```
Comment (sgugger) [inline src/transformers/models/x_clip/modeling_x_clip.py:1355]: Flagging this here. Should be uncommented or removed.

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 6]
```diff
@@ -0,0 +1,1486 @@
...
+
+        vision_outputs = self.vision_model(
+            pixel_values=pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        # TODO remove this assertion (vision pooler output)
+        # assert torch.allclose(vision_outputs.pooler_output[0, :3], torch.tensor([-0.2987, 1.0489, 0.3702]), atol=1e-4)
```
Comment (sgugger) [inline src/transformers/models/x_clip/modeling_x_clip.py:1421]: To be cleaned up?

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 7]
```diff
@@ -0,0 +1,1486 @@
...
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        text_embeds = text_outputs[1]
+        text_embeds = self.text_projection(text_embeds)
+
+        # TODO remove this assertion (text pooler output)
+        # assert torch.allclose(text_embeds[0, :3], torch.tensor([-0.2870, -0.3504, 0.0417]), atol=1e-4)
```
Comment (sgugger) [inline src/transformers/models/x_clip/modeling_x_clip.py:1455]: Same here.

File: src/transformers/models/x_clip/test.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/x_clip/test.py:1]: Should not be added.

File: tests/models/x_clip/test_modeling_x_clip.py [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+    def create_and_check_model(self, config, input_ids, attention_mask, pixel_values):
+        model = XCLIPModel(config).to(torch_device).eval()
+        with torch.no_grad():
+            result = model(input_ids, pixel_values, attention_mask)
+        self.parent.assertEqual(
+            result.logits_per_video.shape,
+            (
+                self.vision_model_tester.batch_size,
+                self.text_model_tester.batch_size,
+            ),
```
Comment (sgugger) [inline tests/models/x_clip/test_modeling_x_clip.py:480]: Fits in one line.

File: tests/models/x_clip/test_modeling_x_clip.py [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+        )
+        self.parent.assertEqual(
+            result.logits_per_text.shape,
+            (
+                self.text_model_tester.batch_size,
+                self.vision_model_tester.batch_size,
+            ),
```
Comment (sgugger) [inline tests/models/x_clip/test_modeling_x_clip.py:487]: Here too.

File: utils/check_config_docstrings.py [hunk 11]
```diff
@@ -40,6 +40,7 @@
 
 
 CONFIG_CLASSES_TO_IGNORE_FOR_DOCSTRING_CHECKPOINT_CHECK = {
+    "XCLIPConfig",
```
Comment (sgugger) [inline utils/check_config_docstrings.py:43]: Let's use alphabetical order here.

File: utils/check_repo.py [hunk 12]
```diff
@@ -125,6 +125,8 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = PRIVATE_MODELS.copy() + [
     # models to ignore for model xxx mapping
+    "XCLIPVisionModel",
```
Comment (sgugger) [inline utils/check_repo.py:128]: Same here ideally (I know the list is messed up already)

**Review by sgugger — COMMENTED** [review]
Thanks for adding this new model! Left a couple of comments.

For the import errors, I think you need to add `"xclip"` to the `SPECIAL_MODEL_TYPE_TO_MODULE_NAME` variable in `configuration_auto.py` since the module name is not the model type `xclip` (with potential `-` replaced by `_`).

**Comment by NielsRogge** [comment]
@sgugger thanks a lot, that solved the issue. There seems to be another (small) issue with run_tests_hub:
```
==================================== ERRORS ====================================
_______________ ERROR collecting tests/utils/test_file_utils.py ________________
tests/utils/test_file_utils.py:26: in <module>
    from transformers import *  # noqa F406
src/transformers/utils/import_utils.py:1021: in __getattr__
    value = getattr(module, name)
src/transformers/utils/import_utils.py:1023: in __getattr__
    raise AttributeError(f"module {self.__name__} has no attribute {name}")
E   AttributeError: module transformers.models.clip has no attribute CLIPProcessor
```
Running `RUN_SLOW=yes pytest tests/utils/test_file_utils.py` passes locally for me.

**Comment by sgugger** [comment]
That would be because you moved `CLIPProcessor` in the non-vision dependent objects in the main init (and rightly so) but did not do the same for the models/clip/__init__.py.

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 13]
```diff
@@ -0,0 +1,1499 @@
...
+        >>> import requests
+        >>> from transformers import CLIPProcessor, XCLIPVisionModel
+
+        >>> model = XCLIPVisionModel.from_pretrained("microsoft/xclip-base-patch32")
+        >>> processor = CLIPProcessor.from_pretrained("microsoft/xclip-base-patch32")
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> inputs = processor(images=image, return_tensors="pt")
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:1076]: This part needs to be updated with `XCLIPProcessor`

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 14]
```diff
@@ -0,0 +1,1499 @@
...
+
+class XCLIPVisionEncoderLayer(nn.Module):
+    """
+    This corresponds to the `CrossFramelAttentionBlock` class in the original implementation.
+    """
+
+    def __init__(self, config: XCLIPConfig):
+        super().__init__()
+        self.num_frames = config.num_frames
+
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:400]:
```suggestion
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 15]
```diff
@@ -0,0 +1,1499 @@
...
+        msg_token = self.message_fc(hidden_states[:, 0, :])
+        msg_token = msg_token.view(batch_size, self.num_frames, hidden_size)
+
+        msg_token = msg_token + self.drop_path(self.message_attn(self.message_ln(msg_token))[0])
+        # add dummy sequence dimension
+        msg_token = msg_token.view(-1, 1, hidden_size)
+
+        hidden_states = torch.cat([hidden_states, msg_token], dim=1)
+
+        residual = hidden_states
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:447]: Just double checking, shouldn't this be `residual = hidden_states.clone()` instead?

It seems lines 449-462 would alter `residual` too.

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 16]
```diff
@@ -0,0 +1,1499 @@
...
+
+        if input_ids is None:
+            raise ValueError("You have to specify either input_ids")
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+
+        hidden_states = self.embeddings(input_ids=input_ids, position_ids=position_ids)
+
+        bsz, seq_len = input_shape
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:776]:
```suggestion
        batch_size, seq_len = input_shape
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 17]
```diff
@@ -0,0 +1,1499 @@
...
+        # X_CLIP's text model uses causal mask, prepare it here.
+        # https://github.com/openai/CLIP/blob/cfcffb90e69f37bf2ff1e988237a0fbe41f33c04/clip/model.py#L324
+        causal_attention_mask = self._build_causal_attention_mask(bsz, seq_len, hidden_states.dtype).to(
+            hidden_states.device
+        )
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:781]:
```suggestion
        causal_attention_mask = self._build_causal_attention_mask(batch_size, seq_len, hidden_states.dtype).to(
            hidden_states.device
        )
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 18]
```diff
@@ -0,0 +1,1499 @@
...
+    ) -> Union[Tuple, BaseModelOutputWithPooling]:
+        r"""
+        Returns:
+
+        Examples:
+
+        ```python
+        >>> from PIL import Image
+        >>> import requests
+        >>> from transformers import CLIPProcessor, XCLIPVisionModel
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:1068]:
```suggestion
        >>> from transformers import XCLIPProcessor, XCLIPVisionModel
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 19]
```diff
@@ -0,0 +1,1499 @@
...
+
+        >>> model = XCLIPVisionModel.from_pretrained("microsoft/xclip-base-patch32")
+        >>> processor = CLIPProcessor.from_pretrained("microsoft/xclip-base-patch32")
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:1071]:
```suggestion
        >>> processor = XCLIPProcessor.from_pretrained("microsoft/xclip-base-patch32")
```

File: src/transformers/models/x_clip/modeling_x_clip.py [hunk 20]
```diff
@@ -0,0 +1,1499 @@
...
+            video_features (`torch.FloatTensor` of shape `(batch_size, output_dim`): The video embeddings obtained by
+            applying the projection layer to the pooled output of [`XCLIPVisionModel`] and
+            [`XCLIPMultiframeIntegrationTransformer`].
+
+        Examples:
+
+        ```python
+        >>> from PIL import Image
+        >>> import requests
+        >>> from transformers import CLIPProcessor, XCLIPModel
```
Comment (alaradirik) [inline src/transformers/models/x_clip/modeling_x_clip.py:1341]:
```suggestion
        >>> from transformers import XCLIPProcessor, XCLIPModel
```

**Review by alaradirik — COMMENTED** [review]
Thank you for adding this! Looks good to me overall, I just left a few comments and questions.

File: tests/models/x_clip/test_modeling_x_clip.py [hunk 21]
```diff
@@ -0,0 +1,673 @@
...
+
+    def test_config(self):
+        self.config_tester.run_common_tests()
+
+    def test_model(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    def test_training(self):
+        pass
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/x_clip/test_modeling_x_clip.py:414]: Yes, X-CLIP does support training, this is just skipping the test for the vision encoder ;) this is also done in CLIP

File: src/transformers/models/x_clip/modeling_x_clip.py (hunk 15 above)
Comment (NielsRogge) [reply src/transformers/models/x_clip/modeling_x_clip.py:446]: Hmm this seems to work fine; is it possible that residual just refers to the original hidden states?

Just did a quick experiment:

```
>>> a = "hello"
>>> b = a
>>> a += "niels"
>>> b
'hello'
```

**Comment by NielsRogge** [comment]
@sgugger and @alaradirik - the PR is ready for merge. Kindly asking for your approval :)

File: src/transformers/models/auto/tokenization_auto.py [hunk 22]
```diff
@@ -253,6 +253,13 @@
             ("wav2vec2", ("Wav2Vec2CTCTokenizer", None)),
             ("wav2vec2-conformer", ("Wav2Vec2CTCTokenizer", None)),
             ("wav2vec2_phoneme", ("Wav2Vec2PhonemeCTCTokenizer", None)),
+            (
+                "xclip",
+                (
+                    "CLIPTokenizer",
+                    "CLIPTokenizerFast" if is_tokenizers_available() else None,
+                ),
+            ),
```
Comment (sgugger) [inline src/transformers/models/auto/tokenization_auto.py:262]:
```suggestion
            ("xclip", ("CLIPTokenizer", "CLIPTokenizerFast" if is_tokenizers_available() else None)),
```
Fits in one line here.

**Review by sgugger — APPROVED** [review]
Looking good, thanks again for adding this model!

**Review by alaradirik — COMMENTED** [review]
Looks good to me! Thanks for adding this

**Pull request merged**